File: cses/06_tree_algorithms/1139_distinct-colors/solution.py
import sys


def main():
    data = sys.stdin.buffer.read().split()
    if not data:
        return
    pos = 0
    n = int(data[pos]); pos += 1

    c = [0] * (n + 1)
    for i in range(1, n + 1):
        c[i] = int(data[pos]); pos += 1

    # Danh sách kề (cây vô hướng, sẽ được định hướng từ gốc 1 khi DFS).
    adj = [[] for _ in range(n + 1)]
    for _ in range(n - 1):
        a = int(data[pos]); b = int(data[pos + 1]); pos += 2
        adj[a].append(b)
        adj[b].append(a)

    # DFS lặp (không đệ quy): tính parent[] và order[] (cha luôn đứng trước con).
    parent = [0] * (n + 1)
    order = []
    visited = bytearray(n + 1)
    stack = [1]
    visited[1] = 1
    while stack:
        u = stack.pop()
        order.append(u)
        for v in adj[u]:
            if not visited[v]:
                visited[v] = 1
                parent[v] = u
                stack.append(v)

    # small-to-large: xử lý order theo thứ tự đảo nên con luôn được duyệt trước cha.
    sets = [None] * (n + 1)
    ans = [0] * (n + 1)

    for u in reversed(order):
        # Thêm màu của u vào set (đã gộp sẵn màu của mọi subtree con).
        s = sets[u]
        if s is None:
            s = {c[u]}
        else:
            s.add(c[u])
        # Kích thước set = số màu phân biệt trong subtree của u.
        ans[u] = len(s)

        # Gộp set của u vào set của cha theo quy tắc small-to-large (nhỏ vào lớn).
        p = parent[u]
        if p:
            sp = sets[p]
            if sp is None:
                sets[p] = s
            elif len(s) > len(sp):
                # set của u lớn hơn: gộp set cha (sp) vào s rồi gán lại cho cha
                s.update(sp)
                sets[p] = s
            else:
                sp.update(s)
        sets[u] = None  # giải phóng bộ nhớ set của u sau khi đã gộp

    sys.stdout.buffer.write((' '.join(map(str, ans[1:n + 1]))).encode())


if __name__ == "__main__":
    main()

File: cses/06_tree_algorithms/1139_distinct-colors/solution.cpp
#include <iostream>
#include <memory>
#include <unordered_set>
#include <vector>

int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    std::size_t n;
    std::cin >> n;

    std::vector<int> color(n + 1);
    for (std::size_t node = 1; node <= n; ++node) {
        std::cin >> color[node];
    }

    std::vector<std::vector<std::size_t>> adjacency(n + 1);
    for (std::size_t edge = 0; edge + 1 < n; ++edge) {
        std::size_t a, b;
        std::cin >> a >> b;
        adjacency[a].push_back(b);
        adjacency[b].push_back(a);
    }

    // DFS lặp (không đệ quy): tính parent[] và order[] sao cho cha luôn đứng trước con.
    std::vector<std::size_t> parent(n + 1, 0);
    std::vector<std::size_t> order;
    order.reserve(n);
    std::vector<std::size_t> stack = {1};
    while (!stack.empty()) {
        std::size_t node = stack.back();
        stack.pop_back();
        order.push_back(node);

        for (std::size_t next : adjacency[node]) {
            if (next == parent[node]) {
                continue;
            }
            parent[next] = node;
            stack.push_back(next);
        }
    }

    std::vector<std::unique_ptr<std::unordered_set<int>>> colors(n + 1);
    std::vector<std::size_t> answer(n + 1);

    // Duyệt order theo thứ tự đảo: con luôn được xử lý trước cha (small-to-large).
    for (auto it = order.rbegin(); it != order.rend(); ++it) {
        std::size_t node = *it;
        // Khởi tạo set của node với màu của chính nó.
        colors[node] = std::make_unique<std::unordered_set<int>>();
        colors[node]->insert(color[node]);

        for (std::size_t next : adjacency[node]) {
            if (parent[next] != node) {
                continue;
            }
            // Luôn gộp set NHỎ vào set LỚN: nếu con lớn hơn thì hoán đổi trước.
            if (colors[node]->size() < colors[next]->size()) {
                colors[node].swap(colors[next]);
            }
            colors[node]->insert(colors[next]->begin(), colors[next]->end());
            colors[next].reset();  // giải phóng set của con đã gộp xong
        }
        // Kích thước set = số màu phân biệt trong subtree của node.
        answer[node] = colors[node]->size();
    }

    for (std::size_t node = 1; node <= n; ++node) {
        if (node > 1) {
            std::cout << ' ';
        }
        std::cout << answer[node];
    }
    std::cout << '\n';
}
